sklearn/metrics/pairwise.py

```diff
@@ -585,6 +586,187 @@ def cosine_distances(X, Y=None):
     return S
 
 
+def gower_distances(X, Y=None, feature_weight=None, categorical_features=None):
+    """Computes the gower distances between X and Y
+
+    Gower is a similarity measure for categorical, boolean and numerical mixed
+    data.
+
+    Parameters
+    ----------
+    X : array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    Y : array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    feature_weight :  array-like, shape (n_features)
+        According the Gower formula, feature_weight is an attribute weight.
+
+    categorical_features: array-like, shape (n_features)
+        Indicates with True/False whether a column is a categorical attribute.
+        This is useful when categorical atributes are represented as integer
+        values. Categorical ordinal attributes are treated as numeric, and
+        must be marked as false.
+
+        Alternatively, the categorical_features array can be represented only
+        with the numerical indexes of the categorical attribtes.
+
+    Returns
+    -------
+    similarities : ndarray, shape (n_samples, n_samples)
+
+    Reference
+    -------
+    Gower, J.C., 1971, A General Coefficient of Similarity and Some of Its
+    Properties.
+
+    Notes
+    ------
+    The non-numeric features, and numeric feature ranges are determined from
+    X and not Y.
+
+    No support for sparse matrices.
+    """
+
+    if issparse(X) or issparse(Y):
+        raise TypeError("Sparse matrices are not supported for gower distance")
+
+    y_none = Y is None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X)
+
+    # array_dtype = X.dtype
+
+    # this if is necessary as strangelly the validator is rejecting numeric
+    # arrays with NaN
+    # if np.issubdtype(X.dtype, np.number) and not np.isfinite(X.sum()) and \
+    #        not np.isfinite(X).all():
+    #    array_dtype = np.object
+
+    X, Y = check_pairwise_arrays(X, Y, dtype=np.object)
+
+    n_rows, n_cols = X.shape
+
+    if categorical_features is None:
+        categorical_features = np.full(n_cols, False)
```

Comment:
The warnings at the bottom of https://travis-ci.org/scikit-learn/scikit-learn/jobs/353703483 suggest this is a problem: in old NumPy, `np.full(n_cols, False)` would not return a boolean array. So use explicit dtype

Reply:
Actually, `np.zeros(n_cols, dtype=bool)` is better.

Reply:
done